components/dashboard/src/projects/ProjectSettings.tsx

```diff
@@ -61,33 +65,52 @@ export default function () {
         }
     }, [team]);
 
-    if (!project) return null;
+    const updateProjectSettings = useCallback(
+        (settings: ProjectSettings) => {
+            if (!project) return;
 
-    const updateProjectSettings = (settings: ProjectSettings) => {
-        if (!project) return;
+            const newSettings = { ...project.settings, ...settings };
+            getGitpodService().server.updateProjectPartial({ id: project.id, settings: newSettings });
+            setProject({ ...project, settings: newSettings });
+        },
+        [project, setProject],
+    );
 
-        const newSettings = { ...project.settings, ...settings };
-        getGitpodService().server.updateProjectPartial({ id: project.id, settings: newSettings });
-        setProject({ ...project, settings: newSettings });
-    };
+    const setWorkspaceClass = useCallback(
+        async (value: string) => {
+            if (!project) {
+                return value;
+            }
+            const before = project.settings?.workspaceClasses?.regular;
+            updateProjectSettings({ workspaceClasses: { ...project.settings?.workspaceClasses, regular: value } });
+            return before;
+        },
+        [project, updateProjectSettings],
+    );
 
-    const setWorkspaceClass = async (value: string) => {
-        if (!project) {
-            return value;
-        }
-        const before = project.settings?.workspaceClasses?.regular;
-        updateProjectSettings({ workspaceClasses: { ...project.settings?.workspaceClasses, regular: value } });
-        return before;
-    };
+    const setWorkspaceClassForPrebuild = useCallback(
+        async (value: string) => {
+            if (!project) {
+                return value;
+            }
+            const before = project.settings?.workspaceClasses?.prebuild;
+            updateProjectSettings({ workspaceClasses: { ...project.settings?.workspaceClasses, prebuild: value } });
+            return before;
+        },
+        [project, updateProjectSettings],
+    );
 
-    const setWorkspaceClassForPrebuild = async (value: string) => {
-        if (!project) {
-            return value;
+    const onProjectRemoved = useCallback(() => {
+        // if there's a current team, navigate to team projects
+        if (team) {
+            history.push(`/t/${team.slug}/projects`);
+        } else {
+            history.push("/projects");
         }
-        const before = project.settings?.workspaceClasses?.prebuild;
-        updateProjectSettings({ workspaceClasses: { ...project.settings?.workspaceClasses, prebuild: value } });
-        return before;
-    };
+    }, [history, team]);
+
+    // TODO: Render a generic error screen for when an entity isn't found
+    if (!project) return null;
```

Comment:
This should live after all hooks, so bumped it down since I added a few useCallback() wrappers on the fns that weren't there before.